docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,192 @@
+# Dead letter queue
+
+## Introduction
+
+In Conduit the concept of a Dead letter queue (DLQ) describes the place where
+messages are sent if they are negatively acknowledged (nacked). This can happen
+for a number of reasons:
+
+- A node in the pipeline experiences an error while processing a message (e.g.
+  processor).
+- The destination connector fails to process a message and sends back a nack.
+- The pipeline is stopped forcefully causing in-flight messages to be nacked.
+
+### Goals
+
+The purpose of a DLQ is to make a pipeline resilient to errors that might
+happen while processing messages.
+
+The design for the DLQ should support the following:
+
+- Allow user to reroute nacked messages to any Conduit connector (builtin or
+  standalone).
+- A message sent to the DLQ should reflect the contents it had just before the
+  error happened. Example: if a message went through 2 processors and failed
+  while being processed by a 3rd processor, the DLQ message should reflect any
+  transforms done by the first 2 processors.
+- Ensure message order is preserved when messages are sent to the DLQ.
+  In other words - the order in which messages were produced by the source
+  connector should be preserved, even if messages are nacked in a different
+  order (this is already assured using
+  a [semaphore](https://github.com/ConduitIO/conduit/blob/b6325584d51ec64bc9086faddcdaa2788a5dfa8f/pkg/pipeline/stream/source_acker.go#L35-L37)).
+- If a message failed to be sent to the DLQ the nack action should fail and
+  stop the pipeline.
+- Allow the user to configure the window and number of messages that need to be
+  nacked to consider the pipeline broken and trigger a stop. Example: if 25 of
+  the last 30 messages are nacked stop the pipeline.
+- Messages pushed to the DLQ should contain metadata describing the error and
```

Comment:
Maybe also add the node in which the failure occurred, unless it's already found somewhere else?

Reply:
This information is currently found in the error message. To me it sounds as good enough given that a human will need to read the error to figure out how to reprocess the messages.

Reply:
> This information is currently found in the error message.

That will probably be good enough :+1: And if we ever need it, it'll be easy to add anyway.

> To me it sounds as good enough given that a human will need to read the error to figure out how to reprocess the messages.

That's true. :+1: I was more for the option where you have a separate field, so you can more easily search the DLQ and figure out if it's a systematic failure of a node (e.g. a bad processor) and so on. But again, I don't think it's such a huge thing in this context, and it's easy to add if we ever figure out we need it.